Criterion 0.4.x trait incompatibility

This is an awesome crate. I love being able to see instruction counts/cpu cycles etc for benches.

It looks like some changes are needed for Criterion 0.4.x

```
error[E0277]: the trait bound `Perf: Measurement` is not satisfied
   --> benches/my_bench.rs:31:18
    |
31  | fn bench(c: &mut Criterion<Perf>) {
    |                  ^^^^^^^^^^^^^^^ the trait `Measurement` is not implemented for `Perf`
    |
    = help: the trait `Measurement` is implemented for `WallTime`
note: required by a bound in `Criterion`
   --> /home/c/.cargo/registry/src/github.com-1ecc6299db9ec823/criterion-0.4.0/src/lib.rs:340:25
    |
340 | pub struct Criterion<M: Measurement = WallTime> {
    |                         ^^^^^^^^^^^ required by this bound in `Criterion`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `crit-perf` due to previous error
warning: build failed, waiting for other jobs to finish...
Error: 'cargo bench' returned an error (exit status: 101); unable to continue.
```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "criterion-perf-events"
-version = "0.2.0"
+version = "0.3.0"
 authors = ["Jens Breitbart <jbreitbart@gmail.com>"]
 edition = "2018"
 description = "Measure perf events for criterion"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -11,7 +11,7 @@ readme = "README.md"
 exclude = [".gitignore", ".vscode/*", ".github/*"]
 
 [dependencies]
-criterion = "0.3.4"
+criterion = "0.4.0"
 perfcnt = "0.8.0"
 
 [[bench]]

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -130,6 +130,7 @@ impl ValueFormatter for PerfFormatter {
     fn format_throughput(&self, throughput: &Throughput, value: f64) -> String {
         match throughput {
             Throughput::Bytes(b) => format!("{:.4} events/byte", value / *b as f64),
+            Throughput::BytesDecimal(b) => format!("{:.4} events/kilobyte", value / *b as f64),
             Throughput::Elements(b) => format!("{:.4} events/element", value / *b as f64),
         }
     }
@@ -150,6 +151,12 @@ impl ValueFormatter for PerfFormatter {
                     *val /= *n as f64;
                 }
                 "events/byte"
+            }
+            Throughput::BytesDecimal(n) => {
+                for val in values {
+                    *val /= *n as f64;
+                }
+                "events/kilobyte"
             }
             Throughput::Elements(n) => {
                 for val in values {

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -129,9 +129,23 @@ impl ValueFormatter for PerfFormatter {
 
     fn format_throughput(&self, throughput: &Throughput, value: f64) -> String {
         match throughput {
-            Throughput::Bytes(b) => format!("{:.4} events/byte", value / *b as f64),
-            Throughput::BytesDecimal(b) => format!("{:.4} events/kilobyte", value / *b as f64),
-            Throughput::Elements(b) => format!("{:.4} events/element", value / *b as f64),
+            Throughput::Bytes(bytes) => format!("{:.4} events/byte", value / *bytes as f64),
+            Throughput::BytesDecimal(bytes) => {
+                let event_per_byte = value / *bytes as f64;
+
+                let (denominator, unit) = if *bytes < 1000 {
+                    (1.0, "events/byte")
+                } else if *bytes < 1000 * 1000 {
+                    (1000.0, "events/kilobyte")
+                } else if *bytes < 1000 * 1000 * 1000 {
+                    (1000.0 * 1000.0, "events/megabyte")
+                } else {
+                    (1000.0 * 1000.0 * 1000.0, "events/gigabyte")
+                };
+
+                format!("{:.4} {}", event_per_byte / denominator, unit)
+            }
+            Throughput::Elements(bytes) => format!("{:.4} events/element", value / *bytes as f64),
         }
     }
 
@@ -146,21 +160,34 @@ impl ValueFormatter for PerfFormatter {
         values: &mut [f64],
     ) -> &'static str {
         match throughput {
-            Throughput::Bytes(n) => {
+            Throughput::Bytes(bytes) => {
                 for val in values {
-                    *val /= *n as f64;
+                    *val /= *bytes as f64;
                 }
                 "events/byte"
             }
-            Throughput::BytesDecimal(n) => {
+            Throughput::BytesDecimal(bytes) => {
+                let bytes_per_second = *bytes;
+                let (denominator, unit) = if bytes_per_second < 1000 {
+                    (1.0, "events/byte")
+                } else if bytes_per_second < 1000 * 1000 {
+                    (1000.0, "events/kilobyte")
+                } else if bytes_per_second < 1000 * 1000 * 1000 {
+                    (1000.0 * 1000.0, "events/megabyte")
+                } else {
+                    (1000.0 * 1000.0 * 1000.0, "events/gigabyte")
+                };
+
                 for val in values {
-                    *val /= *n as f64;
+                    *val /= *bytes as f64;
+                    *val /= denominator;
                 }
-                "events/kilobyte"
+
+                unit
             }
-            Throughput::Elements(n) => {
+            Throughput::Elements(bytes) => {
                 for val in values {
-                    *val /= *n as f64;
+                    *val /= *bytes as f64;
                 }
                 "events/element"
             }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -124,7 +124,7 @@ struct PerfFormatter;
 
 impl ValueFormatter for PerfFormatter {
     fn format_value(&self, value: f64) -> String {
-        format!("{:.4} cycles", value)
+        format!("{value:.4} cycles")
     }
 
     fn format_throughput(&self, throughput: &Throughput, value: f64) -> String {

